### PR TITLE
SHAP Interpretability method implementation

### PR DESCRIPTION
**Pull request by naveenkcb**
Contributor: Naveen Baskaran

Contribution Type: Interpretability method, Tests, Example 

### Description
This PR implements the SHAP (SHapley Additive exPlanations) interpretability method for PyHealth models, enabling users to understand which features contribute most to model predictions. SHAP is based on coalitional game theory and provides theoretically grounded feature importance scores with desirable properties like local accuracy, missingness, and consistency.

### Files to Review
pyhealth/interpret/methods/__init__.py 
pyhealth/interpret/methods/shap.py - Core SHAP method implementation. Suports embedding based attribution, continuous feature support
pyhealth/processors/tensor_processor.py - minor fix to resolve warning message
examples/shap_stagenet_mimic4.py - Example script showing the usage of SHAP method
tests/core/test_shap.py - added comprehensive test cases to test the main class, utility methods and attribution methods.

### Results on mimic4-demo dataset
image

### CODE REVIEW

File: examples/shap_stagenet_mimic4.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+
+# %% Run SHAP on a held-out sample
+print("\n" + "="*80)
+print("Initializing SHAP Explainer")
+print("="*80)
+
+# Initialize SHAP explainer with custom parameters
+shap_explainer = ShapExplainer(
+    model,
+    use_embeddings=True,  # Use embeddings for discrete features
```
Comment (jhnwu3) [inline examples/shap_stagenet_mimic4.py]: I think SHAP should be compatible with discrete tokens like ICD codes here? Correct me if I'm wrong. Will look deeper into understanding the full implementation of SHAP here later when I'm more congitively sound.

Comment (naveenkcb) [reply examples/shap_stagenet_mimic4.py]: I updated the SHAPExplainer class instance creation to pass the just the model and default all other values including "use_embeddings" inside the __init__ method. Yes the SHAP works for ICD codes but will use the embeddings from the input model.

File: pyhealth/interpret/methods/shap.py [hunk 2]
```diff
@@ -0,0 +1,881 @@
...
+            coalition_size: Number of features in the coalition.
+            n_features: Total number of features.
+
+        Returns:
+            Kernel weight as a scalar tensor.
+        """
+        if coalition_size == 0 or coalition_size == n_features:
+            return torch.tensor(1000.0)  # Large weight for edge cases
+
+        comb_val = math.comb(n_features - 1, coalition_size - 1)
```
Comment (jhnwu3) [inline pyhealth/interpret/methods/shap.py]: Wait, isn't it binom(M, |z|) here? Why do we take n_features -1 and coalition-size -1 instead of n_features and coalition_size?

File: pyhealth/interpret/methods/shap.py [hunk 3]
```diff
@@ -0,0 +1,881 @@
...
+        """
+        device = input_emb[key].device
+        batch_size = input_emb[key].shape[0] if input_emb[key].dim() >= 2 else 1
+        n_coalitions = min(2 ** n_features, self.max_coalitions)
+
+        # Storage for coalition sampling
+        coalition_vectors = []
+        coalition_weights = []
+        coalition_preds = []
+
```
Comment (jhnwu3) [inline pyhealth/interpret/methods/shap.py:375]: Can you check if we don't need to add the edge case coalitions specifically (full features and no features) in the prediction set for training the kernel/linear model for predicting shapley values here?

I've linked some captum code examples here:

https://github.com/meta-pytorch/captum/blob/master/captum/attr/_core/kernel_shap.pyhttps://github.com/meta-pytorch/captum/blob/master/captum/attr/_core/lime.py

File: pyhealth/interpret/methods/shap.py (hunk 2 above)
Comment (naveenkcb) [reply pyhealth/interpret/methods/shap.py]: I updated the code to match the equation used in the original SHAP paper in the method _compute_kernel_weight

weight = (M - 1) / (binom(M, |z|) * |z| * (M - |z|))

1. I also added the .rst file as requested.
2. Added "examples/shap_stagenet_mimic4.ipynb" using colab with GPU

File: pyhealth/interpret/methods/shap.py (hunk 3 above)
Comment (naveenkcb) [reply pyhealth/interpret/methods/shap.py:375]: handled the edge cases and updated the code accordingly in the method _compute_kernel_shap

File: pyhealth/interpret/methods/shap.py [hunk 4]
```diff
@@ -0,0 +1,913 @@
...
+                per_input_preds.append(pred)
+
+            coalition_vectors.append(coalition.float())
+            coalition_preds.append(torch.stack(per_input_preds, dim=0))
+            coalition_weights.append(
+                self._compute_kernel_weight(coalition.sum().item(), n_features)
+            )
+
+        # Sample remaining coalitions randomly (excluding edge cases already added)
+        n_random_coalitions = max(0, n_coalitions - 2)
```
Comment (jhnwu3) [inline pyhealth/interpret/methods/shap.py:403]: Hey one last request, as I've been digging deeper into the official shap implementation of kernel regression. They of course do it in numpy, which is mostly so people can interpret random forests and XGBoost.

But, I think we can adopt some of the nice tricks in coalition sampling here:

https://github.com/shap/shap/blob/master/shap/explainers/_kernel.py
Specifically, it seems they do some type of compliment sampling to optimize how much coverage of samples we can sample at a time.

https://github.com/shap/shap/blob/ace49bf463a802f18725a869a969c060a192e3f8/shap/explainers/_kernel.py#L480

Let me know if you need any help with this. It does look a little complicated, but everything else looks good to me.

Comment (naveenkcb) [reply pyhealth/interpret/methods/shap.py:403]: @jhnwu3  based on our discord convo, I am keeping the torch implementation asis to support for StageNet. I did update for the create dataset portions in the unit test and example script. Hope this helps. Please advise if futher changes may be required.

File: docs/api/interpret/pyhealth.interpret.methods.shap.rst
Comment (jhnwu3) [inline docs/api/interpret/pyhealth.interpret.methods.shap.rst:1]: Ah, I just realized, can you add a reference to this here: https://github.com/sunlabuiuc/PyHealth/blob/master/docs/api/interpret.rst 

This way the autodocs knows to add your SHAP Implementation to the front page.

**Review by jhnwu3 — CHANGES_REQUESTED** [review]
Just one last doc change, and then lgtm! Will more robustly test and decide if we need revisions as we interpret the model/look into it deeper. (i.e we might need to compare our implementation with another given some workarounds).

**Comment by naveenkcb** [comment]
@jhnwu3  I added the comment to interpret.rst file and pushed my changes.

**Review by jhnwu3 — CHANGES_REQUESTED** [review]
I wish I could do this haha, but can you delete the .vscode folder here?

**Comment by naveenkcb** [comment]
I removed the .vscode folder. Hope you can merge it now.

**Review by jhnwu3 — APPROVED** [review]
lgtm, thanks for the hard work!
